svix-cli/src/cmds/login.rs

```diff
@@ -45,3 +59,100 @@ pub fn prompt() -> Result<()> {
     );
     Ok(())
 }
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CliStartLoginSessionOut {
+    session_id: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct AuthTokenOut {
+    token: String,
+}
+
+const DASHBOARD_URL: &str = "https://dashboard.svix.com";
+const LOGIN_SERVER_URL: &str = "https://api.svix.com";
+
+const MAX_POLL_TIME: Duration = Duration::from_secs(5 * 60);
+
+pub async fn dashboard_login() -> Result<String> {
+    let client = reqwest::Client::new();
+
+    let start_session = match client
+        .post(format!(
+            "{LOGIN_SERVER_URL}/dashboard/cli/login/session/start"
+        ))
+        .send()
+        .await
+    {
+        Ok(res) => res.json::<CliStartLoginSessionOut>().await?,
+        Err(e) => {
+            eprintln!("Failed to get session ID: {}", e);
+            return Err(anyhow::anyhow!("Failed to get session ID"));
+        }
+    };
+    let session_id = start_session.session_id;
+    let code = &session_id[0..4].to_uppercase();
+
+    let url = format!("{DASHBOARD_URL}/cli/login?sessionId={session_id}&code={code}");
+
+    println!("\nPlease approve the login in your browser, then return here.");
+    println!("Verification code: \x1b[32m{}\x1b[0m\n", code);
+
+    if let Err(e) = open::that(&url) {
+        eprintln!("Failed to open browser: {}", e);
+        println!("Please manually open this URL in your browser: {}", url);
+    }
+
+    println!("Waiting for approval...");
+
+    let poll_url = format!("{LOGIN_SERVER_URL}/dashboard/cli/login/session/complete");
+
+    let start_time = Instant::now();
+
+    while start_time.elapsed() < MAX_POLL_TIME {
+        let response = match client
+            .post(&poll_url)
+            .json(&serde_json::json!({ "sessionId": session_id }))
+            .send()
+            .await
+        {
+            Ok(resp) => resp,
+            Err(e) => {
+                return Err(anyhow::anyhow!(
+                    "Failed to connect to authentication server: {}",
+                    e
+                ));
+            }
+        };
```

Comment:
```suggestion
        let response = client
            .post(&poll_url)
            .json(&serde_json::json!({ "sessionId": session_id }))
            .send()
            .awai
            .context("Failed to connect to authentication server")?;
```

(need to `use anyhow::Context`)

Reply:
ohh cool, I didn't know that. Updated the code to use `Context`